src/Engine/GameEvents/GameEventListener.java

```diff
@@ -0,0 +1,53 @@
+package Engine.GameEvents;
+
+import java.util.HashSet;
+
+import Engine.Combat.BattleSummary;
+import Terrain.Location;
+import Units.Unit;
+
+/**
+ * Subclasses can register to be notified of any supported event types. This class is both the event distributor,
+ * and the extension point for all listeners. Listeners will register to receive events on creation, and should
+ * override those receive functions they care about.
+ * We use the visitor pattern to call the correct event function whenever an event is distributed.
+ */
+public abstract class GameEventListener
+{
+  /** Static list of all event subscribers */
+  private static HashSet<GameEventListener> eventListeners = new HashSet<GameEventListener>();
+
+  /** Pass event along to every listener we still have. */
+  public static void publishEvent(GameEvent event)
+  {
+    for( GameEventListener gel : eventListeners )
+    {
+      // The event will call the appropriate receive method in the listener.
+      event.sendToListener(gel);
+    }
+  }
+
+  /** Sign this listener up to receive events. If a listener registers multiple times, it will still
+   *  receive each notification only once. */
+  public static void registerEventListener(GameEventListener listener)
+  {
+    eventListeners.add(listener);
+  }
+
+  /** Unregister this listener. Call this when a listener is no longer needed, so the JVM knows
+   *  it can clean up the object. */
```

Comment:
Could also use weak references, but ensuring clean code is also good.

Reply:
This is a good thought. We'd still need the `unregisterEventListener` though, since some listeners (e.g. as in Patch's ability) may be unregistered and re-registered instead of being destroyed and re-created.
I suppose we could still use weak references as a backup mechanism, to handle the case where someone forgets to unregister a listener before nulling it. I'm not sure about the behavior of garbage collecting in this case though; could a forgotten listener stick around and affect gameplay until the garbage collector deals with it? If so, it would make it harder to track down this kind of bug, because the problem would just disappear at some indeterminate time.
Of course we could always call `System.gc()` before calling the listeners, but that's exactly the kind of thing we aren't supposed to have to deal with in Java.

Reply:
Some brief testing indicates that once set to null, a listener won't be called, even if it hasn't been garbage-collected yet. I think I'm fine with making this change, and seeing how things go if you think its worthwhile. Thoughts?